[Humphrey] iP

grumphrey opened the pull request:
# Duke: Task Manager 🤖

> "I love task managers!" - Somebody

Download [here!](https://github.com/bananamonkey4655/ip/releases/tag/A-Jar)

## Features [💼](https://emojipedia.org/briefcase/)
- **Add/delete** tasks
- **Mark/unmark** tasks as ~~completed~~
- **Save and load** tasks to/from file
- **Find** tasks containing *keyword*

## Supported tasks 📖
1. Todos
2. Deadlines
3. Events

## Why use Duke?
- [x] easy to use
- [x] great for forgetful people
- [x] sassy robot
- [ ] will take over earth  

The `main` method:
```java
public static void main(String[] args) {
        new Duke("src/data/duke.txt").start();
    }
```

Commit:
Add Gradle support

grumphrey committed:
implement skeletal version of bot

grumphrey committed:
add ability to store text entered by user and display them back when requested

grumphrey committed:
add the ability to mark tasks as done/undone

grumphrey committed:
add support for tracking three types of tasks: todos, deadlines, events

grumphrey committed:
use the input/output redirection technique to semi-automate the testing of duke

grumphrey committed:
handle exceptions

grumphrey committed:
add support for deleting tasks from the list

grumphrey committed:
Add saving and loading of tasks from file

grumphrey committed:
Add dates and times

grumphrey committed:
Merge branch 'branch-Level-7'

grumphrey committed:
Merge branch 'branch-Level-8'

grumphrey committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

grumphrey committed:
Add more OOP

grumphrey committed:
Refactor classes into respective packages

grumphrey committed:
Add unit testing for Todo

grumphrey committed:
Add jar file

grumphrey committed:
Refactor code to follow coding standard

grumphrey committed:
Add javadocs for classes and methods

grumphrey committed:
Add ability to search for tasks containing a matching keyword

grumphrey committed:
Merge branch 'branch-A-CodingStandard'

grumphrey committed:
Merge branch 'branch-A-JavaDoc'

grumphrey committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Task.java

lesterong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+package duke;
+
+import duke.command.*;
```
Should this be separated into explicit imports?

grumphrey replied:
Hey you're right, thanks for spotting that!

lesterong commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.println(botReply);
+    }
+
+    public void welcome() {
```
Should this method be named as a verb instead, e.g. `sayWelcome()` instead?

lesterong commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        String task = tasks.deleteTask(taskNumber);
+        ui.reply("Noted. I've removed this duke.task:\n " + task + "\nNow you have " +
+                tasks.length() + " tasks in the list");
```
Perhaps the `+` operator should be placed on the second line to improve readability. 
```suggestion
        ui.reply("Noted. I've removed this duke.task:\n " + task + "\nNow you have "
                + tasks.length() + " tasks in the list");
```

lesterong commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,49 @@
+        this.sc = new Scanner(System.in);
+    }
+
+    public String readCommand() {
```
Should this method be renamed to better reflect the return value?
```suggestion
    public String getCommand() {
```

lesterong reviewed (commented):
Hey @bananamonkey4655, this looks good to me. I don't see much issues with your variable and method names and code quality. Good job!

grumphrey committed:
Update src/main/java/duke/command/DeleteCommand.java

Improve code readability

Co-authored-by: lesterong <84223259+lesterong@users.noreply.github.com>

grumphrey committed:
Add GUI to task manager using JavaFX

grumphrey committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/duke/command/DeleteCommand.java

jasonchristopher21 commented on src/main/java/duke/manager/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return The number of current tasks.
+     */
+    public int length() {
```
```suggestion
    public int getLength() {
```
Perhaps using a verb as the method name?

grumphrey replied:
Thanks! Good spot

jasonchristopher21 reviewed (commented):
Looking good 👍 
One method name change requested

seanflyyy commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public String execute(TaskList tasks, Storage storage) {
+        tasks.addTask(this.task);
+        String response = "Got it. I've added this duke.task:\n " + this.task +
```
Could consider having the response be nested in a UI method to satisfy the SLAP hard requirement :D

![image](https://user-images.githubusercontent.com/50956270/188150396-abdc2e76-da5c-437a-b97a-66045c4e9ef0.png)

grumphrey replied:
Oh yeah good spot, thanks!

seanflyyy commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
+        String task = tasks.deleteTask(taskNumber);
+        String response = "Noted. I've removed this duke.task:\n " + task + "\nNow you have "
```
Same thing as above comment for AddCommand class. You could consider separating the responses for all your command classes out into the UI class

seanflyyy commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,53 @@
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/DaDuke.png"));
```
Can consider making this final since it is unlikely it will change

seanflyyy commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public String getTask() {
+        String done = this.isDone ? "1" : "0";
+        return String.format("E | %s | %s | %s", done, this.description, this.date);
```
Very smart way of generating the expected output!

seanflyyy commented on src/main/java/duke/manager/Parser.java:
```diff
@@ -0,0 +1,160 @@
+            return new AddCommand(newTask);
+        }
+
+        if (input.startsWith("deadline")) {
```
I like the usage of startWith

seanflyyy reviewed (commented):
Overall really good code! Left quite a few comments that I hope will help!

grumphrey committed:
Test commit

grumphrey committed:
Add assertions to Duke

Hidden bugs may be present in the code.

Using assertions will make sure our code behaves the way we expect, and make our codebase more robust.

Let's add some assertions to a few classes, such as TaskList to ensure the methods of the list behave as we expect.

grumphrey committed:
Improve code quality by extracting code into separate functions in Parser

grumphrey committed:
Add final to variables

grumphrey committed:
Merge pull request #2 from bananamonkey4655/branch-A-CodeQuality

Merge with branch A-CodeQuality

grumphrey committed:
Merge pull request #1 from bananamonkey4655/branch-A-Assertions

Add assertions to Duke

grumphrey committed:
Add reminders feature

grumphrey committed:
Add javadocs for new methods in Deadline

grumphrey committed:
Add javadocs for reminders feature in TaskList

grumphrey committed:
Improve GUI background color and font

grumphrey committed:
Change up bot's personality

grumphrey committed:
Add UI image to docs

grumphrey committed:
Update user guide for task manager bot

grumphrey committed:
Update bot replies

grumphrey committed:
Fix label ellipsis bug
